### PR TITLE
Integration to BigBlueButton

### LINKED ISSUE #465: Integration of BigBlueButton
Thank you very much for this great project.

I would like to suggest to integrate the video conferencing platform BigBlueButton as Jitsi Meet. Jitsi Meet is very easy to use for smaller meetings. BigBlueButton has advantages for larger events, seminars etc..

The API calls are described [here](https://docs.bigbluebutton.org/dev/api.html). For the "create" API call at least custom properties for BBBmeetingID, BBBattendeePW and BBBmoderatorPW are needed. This "create" API call follows a "join" API call together with the fullName and the desired password (selectable joining as moderator or attendee). Best would be, all create parameters could be optional provided as custom properties (BBB[nameofparameter]).

In addition a secret and a baseURL has to be defined on the server itself, which are needed for the checksum calculation and the API call.

Thanks in advance!

### PR DESCRIPTION
**Pull request by fcecagno**
This work is based on #602, refactored and merged to develop.

- A new special zone called BBB can be added to a map (https://github.com/mconf/workadventure-map-startup/blob/master/src/assets/Special_Zones.png);
- Layer must have the following custom properties:
  - `bbbMeeting`: `meetingID` of the BigBlueButton API
  - `meetingName`: `meetingName` of the BigBlueButton API
- Layer might have the following custom properties:
  - `userdata-bbb_auto_share_webcam`: when `true`, webcam will be shared automatically
  - `userdata-bbb_hide_presentation`: when `true`, presentation will be minimized and webcams will use the whole space
  - `userdata-bbb_listen_only_mode`: when `false`, microphone will be enabled for audio and listen only mode won't be used 
  - `userdata-bbb_skip_check_audio`: when `true`, echo test when enabling microphone won't be displayed
  - `userdata-bbb_skip_video_preview`: when `true`, webcam preview won't be displayed before sharing
  - `userdata-bbb_show_participants_on_login`: when `false`, users list won't be displayed
(above custom properties have been enabled on https://github.com/mconf/workadventure-map-startup)

This is how it looks like:

https://user-images.githubusercontent.com/502164/157167960-5eaada2d-1761-4ab8-9669-803861c0159f.mp4

An open test server is available on https://play.workadventure.mconf.dev/_/global/mconf.github.io/workadventure-map-startup/map.json, if anybody is interested to test.

At this moment, all participants are moderators and meetings can be accessed by any user.

fixes #465
closes #602

### CODE REVIEW

File: maps/package.json [hunk 1]
```diff
@@ -44,6 +44,7 @@
     "@typescript-eslint/parser": "^2.26.0",
     "eslint": "^6.8.0",
     "jasmine": "^3.5.0",
+    "generic-type-guard": "^3.2.0",
```
Comment (moufmouf) [inline maps/package.json:47]: I'm pretty sure this package in the "maps" directory is not needed at all.

**Review by moufmouf — COMMENTED** [review]
Hey @fcecagno ,

Thanks a lot for this PR!
I'm having quite an hectic week so I don't expect to have the time to review your PR before next week, but this is definitely cool. I'll keep you posted!

**Comment by fcecagno** [comment]
Hi @moufmouf , thanks!
Just saw a bunch of lint errors on the pipeline, we'll take a closer look and send an update shortly.

**Comment by tabascoeye** [comment]
do you need to modify the BBB server to accept being "embedded" in the iframe?
Or does this work with any open BBB instance with some magic trick?

**Comment by fcecagno** [comment]
Hi @tabascoeye , no modification is needed on BigBlueButton for this.
In the live demo we made available, it's just BigBlueButton server running the latest version installed using https://github.com/bigbluebutton/bbb-install/

**Comment by fcecagno** [comment]
Hi @moufmouf , we believe it looks good for review.
Please let us know if you have any question or need any help.

**Comment by fcecagno** [comment]
Hey @moufmouf , just curious if you had the chance to take a look at this.

File: back/src/Services/SocketManager.ts [hunk 2]
```diff
@@ -614,6 +618,49 @@ export class SocketManager {
         user.socket.write(serverToClientMessage);
     }
 
+    public async handleJoinBBBMeetingMessage(user: User, joinBBBMeetingMessage: JoinBBBMeetingMessage) {
+        const meetingId = joinBBBMeetingMessage.getMeetingid();
+        const meetingName = joinBBBMeetingMessage.getMeetingname();
+
+        const api = BigbluebuttonJs.api(BBB_URL, BBB_SECRET);
```
Comment (moufmouf) [inline back/src/Services/SocketManager.ts:625]: If BBB_URL is empty, we should probably consider:

-  sending an error message to the client (using the ErrorMessage protobuf message)
- logging an error in the pusher logs using console.error

**Comment by moufmouf** [comment]
Hey @fcecagno ,

Thanks a lot for the PR and sorry for taking so much time to review this.

This looks really pretty good!

One thing that is missing is a proper documentation of the feature. Would you mind writing one in a new ".md" file in the "/docs/maps" directory?

File: back/src/Services/SocketManager.ts (hunk 2 above)
Comment (fcecagno) [reply back/src/Services/SocketManager.ts:642]: I believe it has been addressed by this new commit https://github.com/thecodingmachine/workadventure/pull/1948/commits/fd93899381ef7694caa9b59d75ac43cecf8aaee5

**Comment by fcecagno** [comment]
@moufmouf just added the docs, please let me know if this is the correct approach.
I converted the PR to draft just to adjust the meetingID calculation, but early next week it's going to be ready for complete review.

File: contrib/docker/.env.prod.template [hunk 3]
```diff
@@ -58,6 +58,13 @@ JITSI_PRIVATE_MODE=false
 JITSI_ISS=
 SECRET_JITSI_KEY=
 
+#
+# BigBlueButton
+#
+
+BBB_URL: https://test-install.blindsidenetworks.com/bigbluebutton/
+BBB_SECRET: 8cd8ef52e8e101574e400365b55e11a6
```
Comment (moufmouf) [inline contrib/docker/.env.prod.template:66]:
```suggestion
BBB_URL=https://test-install.blindsidenetworks.com/bigbluebutton/
BBB_SECRET=8cd8ef52e8e101574e400365b55e11a6
```

**Comment by moufmouf** [comment]
Hey @fcecagno ,

Thanks a lot!

I started working on the ability to configure the BBB server from the Admin API (so that we can expose the BBB settings also in the SAAS version). I'll probably modify the doc a bit to explain the 2 options when I'm done.

Also, do you think it is ok to commit the test credential of Blindsidenetworks? I was not aware of their BBB solution and don't want to put too much pressure on their environment if a lot of people start installing WA.

**Comment by fcecagno** [comment]
Hi @moufmouf,

Sounds good about the docs, please let me know if you need us to change anything.

Regarding test-install, it's a public server configured by default in all integrations, including [Moodle](https://github.com/blindsidenetworks/moodle-mod_bigbluebuttonbn/blob/v2.4-stable/config-dist.php#L64-L65=) and [Mattermost](https://github.com/blindsidenetworks/mattermost-plugin-bigbluebutton/blob/master/plugin.json#L18-L31=), and it's available exactly to reduce friction for people to try out integrations. I just spoke to @ffdixon and confirm it can be the default server for this integration as well.

**Comment by ffdixon** [comment]
Confirmed!

**Comment by fcecagno** [comment]
Hi @moufmouf , just fixed conflicts for this PR to be mergeable. Would you need any assistance testing it? Please let us know - I'm on Discord as well in case you want to chat about it. We'd love to have this merged in for the next release.

**Pull request merged**